enrol/autoenrol/lib.php

```diff
@@ -155,50 +161,84 @@ public function enrol_allowed($USER, stdClass $instance) {
             }
         }
 
+        if (!$this->check_rule($instance, $user)) {
+            return false;
+        }
+
+        if ($instance->enrolstartdate != 0 and $instance->enrolstartdate > time()) {
+            return false;
+        }
+
+        if ($instance->enrolenddate != 0 and $instance->enrolenddate < time()) {
+            return false;
+        }
+        return true;
+    }
+
+    private function check_rule($instance, $user) {
+        global $CFG;
+
         // Very quick check to see if the user is being filtered.
-        if ($instance->customchar1 != '') {
-            if (!is_object($USER)) {
+        if (!empty($instance->customchar1)) {
+            if (!is_object($user)) {
                 return false;
             }
 
-            if (!isset($USER->auth)) {
-                $USER->auth = '';
-            }
-            if (!isset($USER->department)) {
-                $USER->department = '';
-            }
-            if (!isset($USER->institution)) {
-                $USER->institution = '';
-            }
-            if (!isset($USER->lang)) {
-                $USER->lang = '';
-            }
-            if (!isset($USER->email)) {
-                $USER->email = '';
+            $uservalue = '';
+
+            // Profile field to check.
+            $profileattribute = '';
+            if (isset($instance->customchar3) && ($instance->customchar3 != '-')) {
+                $profileattribute = $instance->customchar3;
+            } else if (isset($instance->customint2)) {
+                $oldfields = array(0 => '', 1 => 'auth', 2 => 'department', 3 => 'institution', 4 => 'lang', 5 => 'email');
+                $profileattribute = $oldfields[$instance->customint2];
             }
 
-            $type = array(1 => $USER->auth, $USER->department, $USER->institution, $USER->lang, $USER->email);
+            $standardfields = array('auth', 'lang', 'department', 'institution', 'address', 'city');
+            if (in_array($profileattribute, $standardfields)) {
+                if (!isset($user->auth)) {
```

Comment:
Hi @bobopinna 

It looks like we'd be able to take care of all this through the API below if we were using Moodle 3.1 as a minimum?

https://docs.moodle.org/dev/User-related_APIs#User_fields_definition

What do you think?